README.md
<a href="https://www.npmjs.com/package/mailhog-awesome">
<img src="https://github.com/derbenoo/mailhog-awesome/raw/master/docs/mailhog-awesome.png?sanitize=true" alt="mailhog-awesome" />
</a>

<p align="center">
NodeJS mailhog client library for email end-to-end testing, written in TypeScript.
</p>

[![npm](https://img.shields.io/npm/v/mailhog-awesome.svg?color=007acc)](https://www.npmjs.com/package/mailhog-awesome) [![GitHub](https://img.shields.io/github/license/derbenoo/mailhog-awesome.svg?color=007acc)](https://github.com/derbenoo/mailhog-awesome/blob/master/LICENSE) [![npm bundle size](https://img.shields.io/bundlephobia/min/mailhog-awesome.svg?color=007acc)](https://www.npmjs.com/package/mailhog-awesome)
[![Snyk Vulnerabilities for npm package](https://img.shields.io/snyk/vulnerabilities/npm/mailhog-awesome.svg)](https://snyk.io/test/npm/mailhog-awesome) [![CircleCI (all branches)](https://img.shields.io/circleci/project/github/derbenoo/mailhog-awesome.svg)](https://circleci.com/gh/derbenoo/mailhog-awesome) 

## :wave: Description

NodeJS client library for [mailhog](https://github.com/mailhog/MailHog), based on the mailhog [npm package](https://www.npmjs.com/package/mailhog). Mailhog is an excellent email testing tool for local development and automated end-to-end testing. While an HTTP API is offerend for implementing tests (e.g. "expect that the reset password email was retrieved"), it is very limited in its capabilities and is not well suited for writing tests. The `mailhog-awesome` package attempts to solve these issues by providing an API that enables developers to write simple and robust email tests. 

The following functionality is provided:

- Inbox system for isolated testing (needed to run tests in parallel)
- Built-in retry logic for email retrieval
- Comprehensive TypeScript typings

## :running: Get Started


```ts
  mailhog = new MailhogClient({
    host: 'localhost',
    port: 8025,
  });

  const emails = await mailhog.getInbox('test1@localhost.org');
```

A simple Jest example test-suite:

```ts
import { createTransport } from 'nodemailer';
import { MailhogClient } from './mailhog';

describe('Example Jest email test-suite', () => {
  let mailhog: MailhogClient;
  const host = 'mailhog';

  beforeAll(async () => {
    mailhog = new MailhogClient({
      host,
      port: 8025,
    });
  });

  it('Retrieve the "reset password" email', async () => {
    // Inbox email only used for this test
    const inbox = 'user1@your-domain.com';

    // Clear the inbox before starting
    await mailhog.clearInbox(inbox);

    // Trigger the "reset password" email via your server's API
    const transport = createTransport({ host, port: 1025, secure: false });
    transport.sendMail({
      from: 'system@your-domain.com',
      to: inbox,
      subject: 'Reset your password',
      html: 'Please follow the link to reset your password.',
    });

    // Get the latest email
    const email = await mailhog.getLastEmail({
      from: 'system@your-domain.com',
      to: inbox,
      subject: 'Reset your password',
    });

    // Expect that a "reset password" email was received
    expect(email).toBeTruthy();

    // Clear inbox after the test has passed
    await mailhog.clearInbox(inbox);
  });
});
```

## :wrench: API


▸ **getInbox**(`email`: string): *Promise‹Email[]›*

Get all emails for an inbox of an email address

**Parameters:**

Name | Type | Description |
------ | ------ | ------ |
`email` | string | email address  |

**Returns:** *Promise‹Email[]›*


---


▸ **clearInbox**(`email`: string): *Promise‹true | MailhogError[]›*

Clear the inbox of an email address

**Parameters:**

Name | Type | Description |
------ | ------ | ------ |
`email` | string | email address  |

**Returns:** *Promise‹true | MailhogError[]›*


---

▸ **getEmails**(`options`: FindEmailOptions): *Promise‹Email[]›*

Get all emails given a set of find options

**Parameters:**

Name | Type | Description |
------ | ------ | ------ |
`options` | FindEmailOptions | email find options  |

**Returns:** *Promise‹Email[]›*


---


▸ **getAllEmails**(): *Promise‹Email[]›*

Get all emails.

**Returns:** *Promise‹Email[]›*



---


▸ **getLastEmail**(`options`: FindEmailOptions): *Promise‹Email | undefined›*

Get the most recent email given the provided find options.

**Parameters:**

Name | Type | Default | Description |
------ | ------ | ------ | ------ |
`options` | FindEmailOptions |  {} | Email find options  |

**Returns:** *Promise‹Email | undefined›*


---

▸ **deleteEmail**(`id`: string): *Promise‹true | MailhogError›*

Delete the email with the given ID.

**Parameters:**

Name | Type | Description |
------ | ------ | ------ |
`id` | string | ID of the email  |

**Returns:** *Promise‹true | MailhogError›*


---

▸ **deleteEmails**(`options`: FindEmailOptions): *Promise‹true | MailhogError[]›*

Delete all emails matching the given find options

**Parameters:**

Name | Type | Description |
------ | ------ | ------ |
`options` | FindEmailOptions | email find options  |

**Returns:** *Promise‹true | MailhogError[]›*


---


▸ **deleteAllEmails**(): *Promise‹true | MailhogError[]›*

Delete all emails.

**Returns:** *Promise‹true | MailhogError[]›*


---

▸ **releaseEmail**(`id`: string, `config`: ReleaseSmtpConfig): *Promise‹true | MailhogError›*

Releases the mail with the given ID using the provided SMTP config.

**Parameters:**

Name | Type | Description |
------ | ------ | ------ |
`id` | string | message ID |
`config` | ReleaseSmtpConfig | SMTP configuration  |

**Returns:** *Promise‹true | MailhogError›*


---

▸ **releaseEmails**(`config`: ReleaseSmtpConfig, `options`: FindEmailOptions): *Promise‹true | MailhogError[]›*

Release all emails that match the given find options using the provided SMTP config.

**Parameters:**

Name | Type | Description |
------ | ------ | ------ |
`config` | ReleaseSmtpConfig | SMTP configuration |
`options` | FindEmailOptions | Email find options  |

**Returns:** *Promise‹true | MailhogError[]›*


---

▸ **releaseAllEmails**(`config`: ReleaseSmtpConfig): *Promise‹true | MailhogError[]›*

Release all emails using the provided SMTP config.

**Parameters:**

Name | Type | Description |
------ | ------ | ------ |
`config` | ReleaseSmtpConfig | SMTP configuration  |

**Returns:** *Promise‹true | MailhogError[]›*


---


▸ **encode**(`str`: string, `encoding`: string, `charset?`: string, `lineLength?`: number): *string*

Encodes a String in the given charset to base64 or quoted-printable encoding.

**Parameters:**

Name | Type | Description |
------ | ------ | ------ |
`str` | string | String to encode |
`encoding` | string | base64/quoted-printable |
`charset?` | string | Charset of the input string (default: 'utf8') |
`lineLength?` | number | Soft line break limit (default: 76)  |

**Returns:** *string*


---

▸ **decode**(`str`: string, `encoding`: string, `charset?`: string): *string*

Decodes a String from the given encoding and outputs it in the given charset.

**Parameters:**

Name | Type | Description |
------ | ------ | ------ |
`str` | string | String to decode |
`encoding` | string | base64/quoted-printable |
`charset?` | string | Charset to use for the output (default: 'utf8')  |

**Returns:** *string*

---

▸ **extractUrls**(`str`: string): *string[]*

Extract all URLs from a string

**Parameters:**

Name | Type | Description |
------ | ------ | ------ |
`str` | string | string to extract all URLs from  |

**Returns:** *string[]*


## :rotating_light: Known Issues

- The "bcc" field is not retrieved by mailhog. 
- Not optimized for performance on large sets of emails: Due to the limited functionality offered by the mailhog web API, the `mailhog-awesome` package has to perform operations that have a sub-optimal runtime complexity, which will be noticeable for larger amounts of emails. As the primary focus of this package is the implementation of end-to-end email tests, this can be easily avoided by properly isolating test suites and clearing the inbox after a test has finished. 

## :pray: Contributing

You are welcome to contribute to the mailhog-awesome GitHub repository! All infos can be found here: [How to contribute](https://github.com/derbenoo/mailhog-awesome/blob/master/CONTRIBUTING.md)

## :book: License

Mailhog-awesome is [MIT licensed](https://github.com/derbenoo/mailhog-awesome/blob/master/LICENSE).


CHANGELOG.md
# CHANGELOG

## 0.1.4

- Export TypeScript types
- Add the `extractUrls` function
- Add default decoding options

## 0.1.3

- Fix TypeScript compilation process

## 0.1.2 

- Add TypeScript definitions to published package

## 0.1.1

- Add package description

## 0.1.0

- Inital version of `mailhog-awesome` released :tada:

libs/mailhog-awesome/src/lib/example.spec.ts
import { createTransport } from 'nodemailer';
import { MailhogClient } from './mailhog';

describe('Example Jest email test-suite', () => {
  let mailhog: MailhogClient;
  const host = 'mailhog';

  beforeAll(async () => {
    mailhog = new MailhogClient({
      host,
      port: 8025,
    });
  });

  it('Retrieve the "reset password" email', async () => {
    // Inbox email only used for this test
    const inbox = 'user1@your-domain.com';

    // Clear the inbox before starting
    await mailhog.clearInbox(inbox);

    // Trigger the "reset password" email via your server's API
    const transport = createTransport({ host, port: 1025, secure: false });
    transport.sendMail({
      from: 'system@your-domain.com',
      to: inbox,
      subject: 'Reset your password',
      html: 'Please follow the link to reset your password.',
    });

    // Get the latest email
    const email = await mailhog.getLastEmail({
      from: 'system@your-domain.com',
      to: inbox,
      subject: 'Reset your password',
    });

    // Expect that a "reset password" email was received
    expect(email).toBeTruthy();

    // Clear inbox after the test has passed
    await mailhog.clearInbox(inbox);
  });
});


libs/mailhog-awesome/src/lib/mailhog.spec.ts
import { createTransport, Transporter, SendMailOptions } from 'nodemailer';
import { MailhogClient } from './mailhog';
import * as delay from 'delay';

describe('Mailhog client end-to-end tests', () => {
  const host = 'mailhog';
  const smtpPort = 1025;
  const webPort = 8025;

  let smtpTransport: Transporter;
  let mailhog: MailhogClient;

  beforeAll(async () => {
    smtpTransport = createTransport({ host, port: smtpPort, secure: false });
    mailhog = new MailhogClient({
      host,
      port: webPort,
      defaults: {
        numRetries: 2,
        retryDelayMs: 200,
      },
    });

    // Clear before starting
    await mailhog.deleteAllEmails();
  });

  afterAll(async () => {
    // Clear before finishing
    await mailhog.deleteAllEmails();
  });

  it('[SENDER-1] Retrieve emails for a sender address', async () => {
    const recipient = 'sender-1-rcp@localhost.noexist';
    const sender = 'sender-1-sender@localhost.noexist';

    const email: SendMailOptions = {
      subject: 'Sender-1 Email subject',
      to: recipient,
      from: sender,
      html: 'does not matter',
    };

    await smtpTransport.sendMail(email);

    const emails = await mailhog.getEmails({ from: sender });
    expect(emails.length).toBe(1);
    expect(emails[0]).toMatchObject(email);
  });

  it('[INBOX-1] Retrieve an email via inbox', async () => {
    const recipient = 'inbox-1-rcp@localhost.noexist';
    const sender = 'inbox-1-sender@localhost.noexist';

    const email: SendMailOptions = {
      subject: 'INBOX-1 Email subject',
      to: recipient,
      from: sender,
      html: 'does not matter',
    };

    await smtpTransport.sendMail(email);

    const emails = await mailhog.getInbox(recipient);
    expect(emails.length).toBe(1);
    expect(emails[0]).toMatchObject(email);
  });

  it('[INBOX-2] Retrieve filtered emails for inbox', async () => {
    const recipient = 'inbox-2-rcp@localhost.noexist';

    const emailOptions: SendMailOptions[] = [
      {
        subject: 'INBOX-2 Email 1',
        to: recipient,
        from: 'inbox-2-sender-1@localhost.noexist',
        html: 'INBOX-2 Body 1',
      },
      {
        subject: 'INBOX-2 Email 2',
        to: recipient,
        cc: 'inbox-2-cc-1@localhost.noexist',
        from: 'inbox-2-sender-2@localhost.noexist',
        html: 'INBOX-2 Body 2',
      },
      {
        subject: 'INBOX-2 Email 3',
        to: recipient,
        cc: 'inbox-2-cc-1@localhost.noexist',
        bcc: 'inbox-2-bcc-1@localhost.noexist',
        from: 'inbox-2-sender-1@localhost.noexist',
        html: 'INBOX-2 Body 3',
      },
      {
        subject: 'INBOX-2 Email 4',
        to: recipient,
        bcc: 'inbox-2-bcc-2@localhost.noexist',
        from: 'inbox-2-sender-2@localhost.noexist',
        html: 'INBOX-2 Body 4',
      },
    ];

    // Send all emails
    await Promise.all(emailOptions.map(email => smtpTransport.sendMail(email)));

    // Check CC filter
    const ccEmails = await mailhog.getEmails({ to: recipient, cc: 'inbox-2-cc-1@localhost.noexist' });
    expect(ccEmails.length).toBe(2);
    expect(ccEmails.map(email => email.subject).sort()).toEqual(['INBOX-2 Email 2', 'INBOX-2 Email 3']);

    // Check before filter
    const beforeEmails1 = await mailhog.getEmails({ to: recipient, before: new Date(Date.now() + 20000) });
    expect(beforeEmails1.length).toBe(4);
    const beforeEmails2 = await mailhog.getEmails({ to: recipient, before: new Date(Date.now() - 20000) });
    expect(beforeEmails2.length).toBe(0);

    // Check after filter
    const afterEmails1 = await mailhog.getEmails({ to: recipient, after: new Date(Date.now() - 20000) });
    expect(afterEmails1.length).toBe(4);
    const afterEmails2 = await mailhog.getEmails({ to: recipient, after: new Date(Date.now() + 20000) });
    expect(afterEmails2.length).toBe(0);
  });

  it('[DELAY-1] Receive a delayed email', async () => {
    const recipient = 'delay-1-rcp@localhost.noexist';
    const sender = 'delay-1-sender@localhost.noexist';

    // Send a delayed email
    setTimeout(() => {
      smtpTransport.sendMail({
        subject: 'Mail delayed by 500ms',
        to: recipient,
        from: sender,
        html: 'does not matter',
      });
    }, 500);

    // Get email
    const emails = await mailhog.getEmails({ to: recipient, retryDelayMs: 500, numRetries: 5 });
    expect(emails.length).toBe(1);
  });

  it('[DELAY-2] Give up email retrieval', async () => {
    const recipient = 'delay-2-rcp@localhost.noexist';

    await mailhog.clearInbox(recipient);
    // Get email
    const emails = await mailhog.getInbox(recipient);
    expect(emails.length).toBe(0);
  });

  it('[SEARCH-1] Search for emails in inbox', async () => {
    const recipient = 'search-1-rcp@localhost.noexist';
    const sender = 'search-1-sender@localhost.noexist';

    const email: SendMailOptions = {
      subject: 'INBOX-1 Email subject',
      to: recipient,
      from: sender,
      text: 'NOT-FOUND',
    };

    await Promise.all(
      ['NO SEARSTR', 'YES SEARCH-STR', 'NO SEARCH-ST', 'YES SEAR SEARCH-STR YES', 'SEARCH-STR'].map(
        async (text: string) => {
          await smtpTransport.sendMail({ ...email, text });
        }
      )
    );

    const emails = await mailhog.getEmails({ to: recipient, body: 'SEARCH-STR' });
    expect(emails.length).toBe(3);
  });

  it('[LAST-1] Get the last email', async () => {
    const recipient = 'last-1-rcp@localhost.noexist';
    const sender = 'last-1-sender@localhost.noexist';

    const email1: SendMailOptions = {
      subject: 'LAST-1 Email 1',
      to: recipient,
      from: sender,
      html: 'does not matter',
    };
    await smtpTransport.sendMail(email1);

    await delay(250);

    const email2: SendMailOptions = {
      subject: 'LAST-1 Email 2',
      to: recipient,
      from: sender,
      html: 'does not matter',
    };
    await smtpTransport.sendMail(email2);

    // Get email
    const email = await mailhog.getLastEmail({ to: recipient });
    expect(email).toBeTruthy();
    expect(email.subject).toBe('LAST-1 Email 2');
  });
});


libs/mailhog-awesome/src/lib/mailhog.types.ts
// Types for https://www.npmjs.com/package/mailhog

import { IncomingMessage } from 'http';

/** Mailhog email attachment */
export interface EmailAttachment {
  /** Filename */
  name: string;
  /** Content-Type */
  type: string;
  /** Content-Transfer-Encoding */
  encoding: string;
  /** Encoded content */
  Body: string;
}

/** Mailhog email */
export interface Email {
  /** Mail ID */
  ID: string;
  /** Decoded mail text content */
  text: string;
  /** Decoded mail HTML content */
  html: string;
  /** Decoded mail Subject header */
  subject: string;
  /** Decoded mail From header */
  from: string;
  /** Decoded mail To header */
  to: string;
  /** Decoded mail Cc header */
  cc: string;
  /** Decoded mail Bcc header */
  bcc: string;
  /** Decoded mail Reply-To header */
  replyTo: string;
  /** Mail Date header */
  date: Date;
  /** Mail Delivery-Date header */
  deliveryDate: Date;
  /** List of mail attachments */
  attachments: EmailAttachment[];
}

export interface GetEmailsRsp {
  /** Number of results available */
  total: number;
  /** Number of results returned */
  count: number;
  /** Offset for the range of results returned */
  start: number;
  /** List of mail object items */
  items: Email[];
}

export interface MailhogNodeOptions {
  /** API protocol */
  protocol?: string;
  /** API host */
  host?: string;
  /** API port */
  port?: number;
  /** API basic authentication */
  auth?: string;
  /** API base path */
  basePath?: string;
}

export interface ReleaseSmtpConfig {
  /** SMTP host */
  host: string;
  /** SMTP port */
  port: string;
  /** recipient email */
  email: string;
  /** SMTP username */
  username?: string;
  /** SMTP password */
  password?: string;
  /** SMTP auth type (PLAIN or CRAM-MD5) */
  mechanism?: 'PLAIN' | 'CRAM-MD5';
}

export interface MailhogNodeClient {
  /** Mailhog Node options object*/
  options: MailhogNodeOptions;

  /**
   * Retrieves a list of mail objects, sorted from latest to earliest.
   * @param start defines the messages query offset (default: 0)
   * @param limit defines the max number of results (default: 50)
   */
  messages(start?: number, limit?: number): Promise<GetEmailsRsp>;

  /**
   * Retrieves a list of mail objects for the given query, sorted from latest to earliest.
   * @param query search query
   * @param kind query kind (default: 'containing')
   * @param start defines the search query offset (default: 0)
   * @param limit 	defines the max number of results (default: 50)
   */
  search(query: string, kind?: 'from' | 'to' | 'containing', start?: number, limit?: number): Promise<GetEmailsRsp>;

  /**
   * Retrieves the latest mail object sent from the given address.
   * @param query from address
   */
  latestFrom(query: string): Promise<GetEmailsRsp>;

  /**
   * Retrieves the latest mail object sent to the given address.
   * @param query to address
   */
  latestTo(query: string): Promise<GetEmailsRsp>;

  /**
   * Retrieves the latest mail object containing the given query.
   * @param query search query
   */
  latestContaining(query: string): Promise<GetEmailsRsp>;

  /**
   * Releases the mail with the given ID using the provided SMTP config.
   * @param id message ID
   * @param config SMTP configuration
   */
  releaseMessage(id: string, config: ReleaseSmtpConfig): Promise<IncomingMessage>;

  /**
   * Deletes the mail with the given ID from MailHog.
   * @param id message ID
   */
  deleteMessage(id: string): Promise<IncomingMessage>;

  /**
   * Deletes all mails stored in MailHog.
   */
  deleteAll(): Promise<IncomingMessage>;

  /**
   * Encodes a string in the given charset to base64 or quoted-printable encoding.
   * @param str string to encode
   * @param encoding base64/quoted-printable
   * @param charset Charset of the input string (default: 'utf8')
   * @param lineLength Soft line break limit (default: 76)
   */
  encode(str: string, encoding: string, charset?: string, lineLength?: number): string;

  /**
   * Decodes a string from the given encoding and outputs it in the given charset.
   * @param str string to decode
   * @param encoding base64/quoted-printable
   * @param charset Charset to use for the output (default: 'utf8')
   */
  decode(str: string, encoding: string, charset?: string): string;
}


libs/mailhog-awesome/src/lib/mailhog-node.types.ts
// Types for https://www.npmjs.com/package/mailhog

import { IncomingMessage } from 'http';

export interface IMailhogNodeAttachment {
  /** Filename */
  name: string;
  /** Content-Type */
  type: string;
  /** Content-Transfer-Encoding */
  encoding: string;
  /** Encoded content */
  Body: string;
}

export interface IMailhogNodeItem {
  /** Mail ID */
  ID: string;
  /** Decoded mail text content */
  text: string;
  /** Decoded mail HTML content */
  html: string;
  /** Decoded mail Subject header */
  subject: string;
  /** Decoded mail From header */
  from: string;
  /** Decoded mail To header */
  to: string;
  /** Decoded mail Cc header */
  cc: string;
  /** Decoded mail Bcc header */
  bcc: string;
  /** Decoded mail Reply-To header */
  replyTo: string;
  /** Mail Date header */
  date: Date;
  /** Mail Delivery-Date header */
  deliveryDate: Date;
  /** List of mail attachments */
  attachments: IMailhogNodeAttachment[];
}

export interface IMailhogNodeMessagesRsp {
  /** Number of results available */
  total: number;
  /** Number of results returned */
  count: number;
  /** Offset for the range of results returned */
  start: number;
  /** List of mail object items */
  items: IMailhogNodeItem[];
}

export interface IMailhogNodeOptions {
  /** API protocol */
  protocol?: string;
  /** API host */
  host?: string;
  /** API port */
  port?: number;
  /** API basic authentication */
  auth?: string;
  /** API base path */
  basePath?: string;
}

export interface IReleaseSmtpConfig {
  /** SMTP host */
  host: String;
  /** SMTP port */
  port: String;
  /** recipient email */
  email: String;
  /** SMTP username */
  username?: String;
  /** SMTP password */
  password?: String;
  /** SMTP auth type (PLAIN or CRAM-MD5) */
  mechanism?: 'PLAIN' | 'CRAM-MD5';
}

export interface IMailhogNodeClient {
  /** Mailhog Node options object*/
  options: IMailhogNodeOptions;

  /**
   * Retrieves a list of mail objects, sorted from latest to earliest.
   * @param start defines the messages query offset (default: 0)
   * @param limit defines the max number of results (default: 50)
   */
  messages(start?: number, limit?: number): Promise<IMailhogNodeMessagesRsp>;

  /**
   * Retrieves a list of mail objects for the given query, sorted from latest to earliest.
   * @param query search query
   * @param kind query kind (default: 'containing')
   * @param start defines the search query offset (default: 0)
   * @param limit 	defines the max number of results (default: 50)
   */
  search(
    query: string,
    kind?: 'from' | 'to' | 'containing',
    start?: number,
    limit?: number
  ): Promise<IMailhogNodeMessagesRsp>;

  /**
   * Retrieves the latest mail object sent from the given address.
   * @param query from address
   */
  latestFrom(query: string): Promise<IMailhogNodeMessagesRsp>;

  /**
   * Retrieves the latest mail object sent to the given address.
   * @param query to address
   */
  latestTo(query: string): Promise<IMailhogNodeMessagesRsp>;

  /**
   * Retrieves the latest mail object containing the given query.
   * @param query search query
   */
  latestContaining(query: string): Promise<IMailhogNodeMessagesRsp>;

  /**
   * Releases the mail with the given ID using the provided SMTP config.
   * @param id message ID
   * @param config SMTP configuration
   */
  releaseMessage(
    id: number,
    config: IReleaseSmtpConfig
  ): Promise<IncomingMessage>;

  /**
   * Deletes the mail with the given ID from MailHog.
   * @param id message ID
   */
  deleteMessage(id: number): Promise<IncomingMessage>;

  /**
   * Deletes all mails stored in MailHog.
   */
  deleteAll(): Promise<IncomingMessage>;

  /**
   * Encodes a String in the given charset to base64 or quoted-printable encoding.
   * @param str String to encode
   * @param encoding base64/quoted-printable
   * @param charset Charset of the input string (default: 'utf8')
   * @param lineLength Soft line break limit (default: 76)
   */
  encode(
    str: string,
    encoding: string,
    charset?: string,
    lineLength?: number
  ): string;

  /**
   * Decodes a String from the given encoding and outputs it in the given charset.
   * @param str String to decode
   * @param encoding base64/quoted-printable
   * @param charset Charset to use for the output (default: 'utf8')
   */
  decode(str: string, encoding: string, charset?: string): string;
}


libs/mailhog-awesome/src/lib/mailhog.ts
import { MailhogNodeClient, MailhogNodeOptions, Email, ReleaseSmtpConfig } from './mailhog.types';
import * as mailhog from 'mailhog';
import * as delay from 'delay';
import { IncomingMessage } from 'http';

const HTTP_SUCCESS = 200;

interface DecodingOptions {
  encoding: 'base64' | 'quoted-printable';
  charset?: string;
}

export interface FindEmailOptions {
  /** Only return emails with this exact "from" field */
  from?: string;
  /** Only return emails with this exact "to" field */
  to?: string;
  /** Only return emails with this exact "cc" field */
  cc?: string;
  /** Only return emails where the subject contains the given string */
  subject?: string;
  /** Only return emails where the body contains the given string */
  body?: string;
  /** Only return emails which where received before the given date */
  before?: Date;
  /** Only return emails which where received after the given date */
  after?: Date;
  /** Number of retries when fetching emails before giving up (default: 1)*/
  numRetries?: number;
  /** Delay between email fetching retries (default: 500ms)*/
  retryDelayMs?: number;
  /** Decoding options for email subject, text and html properties */
  decode?: DecodingOptions;
}

type DefaultOptions = Pick<FindEmailOptions, 'numRetries' | 'retryDelayMs' | 'decode'>;

export interface MailhogOptions extends MailhogNodeOptions {
  defaults?: DefaultOptions;
}

export interface MailhogError {
  reason: string;
  rsp: IncomingMessage;
}

export class MailhogClient {
  private readonly limit = 5000;
  private readonly defaults: DefaultOptions;
  private readonly mailhog: MailhogNodeClient;

  constructor(options: MailhogOptions) {
    this.defaults = options.defaults || {};
    this.mailhog = mailhog(options);
  }

  /**
   * Get all emails given a set of find options
   * @param options email find options
   */
  async getEmails(options: FindEmailOptions): Promise<Email[]> {
    const { after, before, body, cc, from, subject, to } = options;
    const numRetries = options.numRetries || this.defaults.numRetries || 1;
    const retryDelayMs = options.retryDelayMs || this.defaults.retryDelayMs || 500;
    const decode = options.decode || this.defaults.decode;

    for (let i = 0; i <= numRetries; i++) {
      let emails: Email[] = [];

      if (body) {
        const { items } = await this.mailhog.search(body, 'containing', 0, this.limit);
        emails = items;
      } else {
        const { items } = await this.mailhog.messages(0, this.limit);
        emails = items;
      }

      // Filter by find options
      emails = emails.filter(email => {
        if (from && email.from !== from) {
          return false;
        }
        if (to && email.to !== to) {
          return false;
        }
        if (cc && email.cc !== cc) {
          return false;
        }
        if (before && email.date.getTime() >= before.getTime()) {
          return false;
        }
        if (after && email.date.getTime() <= after.getTime()) {
          return false;
        }
        if (subject && !email.subject.includes(subject)) {
          return false;
        }
        return true;
      });

      if (emails.length > 0) {
        if (decode) {
          emails.forEach(email => {
            email.subject = this.decode(email.subject, decode.encoding, decode.charset);
            email.text = this.decode(email.text, decode.encoding, decode.charset);
            email.html = this.decode(email.html, decode.encoding, decode.charset);
          });
        }
        return emails;
      }

      await delay(retryDelayMs);
    }

    // No messages found
    return [];
  }

  /**
   * Get all emails.
   */
  async getAllEmails(): Promise<Email[]> {
    return this.getEmails({});
  }

  /**
   * Get the most recent email given the provided find options.
   * @param options Email find options
   */
  async getLastEmail(options: FindEmailOptions = {}): Promise<Email | undefined> {
    const emails = await this.getEmails(options);
    return emails[0];
  }

  /**
   * Delete the email with the given ID.
   * @param id ID of the email
   */
  async deleteEmail(id: string): Promise<true | MailhogError> {
    const rsp = await this.mailhog.deleteMessage(id);
    if (rsp.statusCode !== HTTP_SUCCESS) {
      return { rsp, reason: `Could not delete email with ID ${id}!` };
    }
    return true;
  }

  /**
   * Delete all emails matching the given find options
   * @param options email find options
   */
  async deleteEmails(options: FindEmailOptions): Promise<true | MailhogError[]> {
    const emails = await this.getEmails(options);
    const results = await Promise.all(emails.map(async email => this.deleteEmail(email.ID)));

    const errors = results.filter(result => result !== true) as MailhogError[];
    if (errors.length > 0) {
      return errors;
    }

    return true;
  }

  /**
   * Delete all emails.
   */
  async deleteAllEmails(): Promise<true | MailhogError[]> {
    return this.deleteEmails({});
  }

  /**
   * Releases the mail with the given ID using the provided SMTP config.
   * @param id message ID
   * @param config SMTP configuration
   */
  async releaseEmail(id: string, config: ReleaseSmtpConfig): Promise<true | MailhogError> {
    const rsp = await this.mailhog.releaseMessage(id, config);

    if (rsp.statusCode !== HTTP_SUCCESS) {
      return { rsp, reason: `Could not release email with ID ${id}!` };
    }
    return true;
  }

  /**
   * Release all emails that match the given find options using the provided SMTP config.
   * @param config SMTP configuration
   * @param options Email find options
   */
  async releaseEmails(config: ReleaseSmtpConfig, options: FindEmailOptions): Promise<true | MailhogError[]> {
    const emails = await this.getEmails(options);
    const results = await Promise.all(emails.map(email => this.releaseEmail(email.ID, config)));

    const errors = results.filter(result => result !== true) as MailhogError[];
    if (errors.length > 0) {
      return errors;
    }

    return true;
  }

  /**
   * Release all emails using the provided SMTP config.
   * @param config SMTP configuration
   */
  async releaseAllEmails(config: ReleaseSmtpConfig): Promise<true | MailhogError[]> {
    return this.releaseEmails(config, {});
  }

  /**
   * Get all emails for an inbox of an email address
   * @param email email address
   */
  async getInbox(email: string): Promise<Email[]> {
    return this.getEmails({ to: email });
  }

  /**
   * Clear the inbox of an email address
   * @param email email address
   */
  async clearInbox(email: string): Promise<true | MailhogError[]> {
    return this.deleteEmails({ to: email });
  }

  /**
   * Encodes a String in the given charset to base64 or quoted-printable encoding.
   * @param str String to encode
   * @param encoding base64/quoted-printable
   * @param charset Charset of the input string (default: 'utf8')
   * @param lineLength Soft line break limit (default: 76)
   */
  encode(str: string, encoding: 'base64' | 'quoted-printable', charset?: string, lineLength?: number): string {
    return this.mailhog.encode(str, encoding, charset, lineLength);
  }

  /**
   * Decodes a String from the given encoding and outputs it in the given charset.
   * @param str String to decode
   * @param encoding base64/quoted-printable
   * @param charset Charset to use for the output (default: 'utf8')
   */
  decode(str: string, encoding: 'base64' | 'quoted-printable', charset?: string): string {
    return this.mailhog.decode(str, encoding, charset);
  }

  /**
   * Extract all URLs from a string
   * @param str string to extract all URLs from
   */
  extractUrls(str: string): string[] {
    return str.match(
      /(http(s)?:\/\/.)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)/
    );
  }
}
